toolchain/check/context.cpp

```diff
@@ -505,6 +532,102 @@ auto Context::ImplicitAsRequired(Parse::Node parse_node, SemIR::NodeId value_id,
   return output_value_id;
 }
 
+static auto CanConvertTupleToArray(Context& context, SemIR::Node tuple_type,
+                                   SemIR::Node array_type) -> bool {
+  auto [array_bound_id, element_type_id] = array_type.GetAsArrayType();
+  auto tuple_elem_types_id = tuple_type.GetAsTupleType();
+  const auto& tuple_elem_types =
+      context.semantics_ir().GetTypeBlock(tuple_elem_types_id);
+
+  if (tuple_elem_types.size() !=
+      context.semantics_ir().GetArrayBoundValue(array_bound_id)) {
+    return false;
+  }
+
+  // TODO: We should do a proper convertibility check here for each
+  // tuple element.
+  return llvm::all_of(tuple_elem_types,
+                      [element_type_id = element_type_id](auto type_id) {
+                        return type_id == element_type_id;
+                      });
```

Comment:
I see this in the old code, although it strikes me it should be simple to call `ImplicitAsImpl` for this (pass `tuple_value` and iterate values, instead of `tuple_type`). Though, that may suggest this function should be a member instead of a static function.

Maybe not worth it at the moment since it was like this before, but would particularly help code that includes builtin errors (e.g., `var x: [i32; 1] = (ProduceError(),);` would be "compatible").

Reply:
Obsolete after #3208.